discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -0,0 +1,121 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{document::Document, engine::GenericError};
+use url::Url;
+use xayn_discovery_engine_providers::Article;
+
+pub(crate) trait ArticleFilter {
+    fn apply(current: &[Document], articles: Vec<Article>) -> Result<Vec<Article>, GenericError>;
+}
+
+struct DuplicateFilter;
+impl ArticleFilter for DuplicateFilter {
+    fn apply(
+        current: &[Document],
+        mut articles: Vec<Article>,
+    ) -> Result<Vec<Article>, GenericError> {
+        let urls = current
+            .iter()
+            .map(|a| a.resource.url.as_str())
+            .collect::<Vec<_>>();
+        let filtered = articles
+            .drain(..)
+            .filter(|article| !urls.contains(&article.link.as_str()))
+            .collect();
+        Ok(filtered)
+    }
+}
+
+struct MalformedImageUrlFilter;
+impl ArticleFilter for MalformedImageUrlFilter {
+    fn apply(
+        _current: &[Document],
+        mut articles: Vec<Article>,
+    ) -> Result<Vec<Article>, GenericError> {
+        let filtered = articles
+            .drain(..)
+            .filter(|article| Url::parse(&article.media).is_ok())
+            .collect();
+        Ok(filtered)
+    }
+}
+
+pub(crate) struct CommonFilter;
+impl ArticleFilter for CommonFilter {
+    fn apply(current: &[Document], articles: Vec<Article>) -> Result<Vec<Article>, GenericError> {
+        DuplicateFilter::apply(current, articles)
+            .and_then(|articles| MalformedImageUrlFilter::apply(current, articles))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{
+        document::{document_from_article, Document},
+        stack::filters::CommonFilter,
+    };
+    use xayn_discovery_engine_providers::Article;
+
+    #[test]
+    fn test_duplicate_filter() {
+        let valid_articles: Vec<Article> =
+            serde_json::from_str(include_str!("../../test-fixtures/articles-valid.json")).unwrap();
+
+        let documents = valid_articles.as_slice()[0..2]
+            .iter()
+            .map(|article| {
+                let doc = Document::default();
+                document_from_article(article.clone(), doc.stack_id, doc.smbert_embedding).unwrap()
+            })
+            .collect::<Vec<_>>();
+
+        let result = CommonFilter::apply(documents.as_slice(), valid_articles).unwrap();
+        let titles = result.iter().map(|a| &a.title).collect::<Vec<_>>();
+
+        assert_eq!(titles.as_slice(), [
```

Comment:
no need for the additional referencing, as vectors can be compared with arrays via the macro directly.
```suggestion
        assert_eq!(titles, [
```